[5.1] Updated the test to confirm the url action generator bug.

rigor789 opened the pull request:
By default laravel uses the root namespace of "App\Http\Controllers", to have a controller in another namespace we have to use "\My\Other\Namespace\Controller@foo".

The test wasn't adding the extra backslash, thus giving a false positive result.

See https://github.com/laravel/framework/issues/9206 for more details!

GrahamCampbell commented:
I don't think this is a bug. :(

crynobone commented on tests/Routing/RoutingUrlGeneratorTest.php:
```diff
@@ -128,7 +128,7 @@ public function testControllerRoutesWithADefaultNamespace()
         $routes->add($route);
 
-        $route = new Illuminate\Routing\Route(['GET'], 'something/else', ['controller' => 'something\foo@bar']);
+        $route = new Illuminate\Routing\Route(['GET'], 'something/else', ['controller' => '\something\foo@bar']);
```
The original test should be correct because a few line above there `$url->setRootControllerNamespace('namespace');` which set the root namespace to `namespace`, and due to this `something` namespace would include the leading backslash.

rigor789 replied:
When calling ```$url->action('foo@bar')``` it appends the namespace to match ```namespace\foo@bar```
If you add a ```\``` it skips adding the root namespace since you don't want it.
If ```controller => '\something\foo@bar'``` is set, and you try to use
```$url->action('\something\foo@bar')``` It should match, but since it trims off all of the ```\```, it doesn't match: ```something\foo@bar !=  \something\foo@bar```

crynobone replied:
@rigor789 Still, you can just simply change a working testcase to fit your need, as my argument stand, this is to test registering any route (other than in the root namespace) doesn't required `\` prefixed, and it is to that that you need to include `\` in `action()` other than the root namespace. 

That what @taylorotwell wanted/intended to ensure happen in the code.

crynobone replied:
The original structure is an original implementation since `v5.0.0` release

https://github.com/laravel/framework/blob/v5.0.0/src/Illuminate/Routing/UrlGenerator.php#L522-L534
https://github.com/laravel/framework/blob/v5.0.0/tests/Routing/RoutingUrlGeneratorTest.php#L118-L138

rigor789 replied:
The testcase wasn't correct in the first place. Can you define "App\Test\Foo@bar" as a route in laravel, and have it working correctly?

It will look for ```App\Http\Controller\App\Test\Foo@bar``` and you will get a class not found exception.

crynobone replied:
> Can you define "App\Test\Foo@bar" as a route in laravel, and have it working correctly?

How did you do it in order to make it not working, there many packages in Laravel 5+ that uses routes (obviously outside of the root namespace) and working fine.

```php
<?php namespace App\Test;

use Illuminate\Routing\Controller;

class Foo extends Controller
{
    public function bar()
    {
        return action('\App\Test\Foo@bar');
    }
}
```

```php
<?php

namespace App\Providers;

use Illuminate\Routing\Router;
use Illuminate\Foundation\Support\Providers\RouteServiceProvider as ServiceProvider;

class RouteServiceProvider extends ServiceProvider
{
    /**
     * This namespace is applied to the controller routes in your routes file.
     *
     * In addition, it is set as the URL generator's root namespace.
     *
     * @var string
     */
    protected $namespace = 'App\Http\Controllers';

    /**
     * Define your route model bindings, pattern filters, etc.
     *
     * @param  \Illuminate\Routing\Router  $router
     * @return void
     */
    public function boot(Router $router)
    {
        //

        parent::boot($router);
    }

    /**
     * Define the routes for the application.
     *
     * @param  \Illuminate\Routing\Router  $router
     * @return void
     */
    public function map(Router $router)
    {
        $router->group(['namespace' => $this->namespace], function ($router) {
            require app_path('Http/routes.php');
        });

        $router->get('foo', 'App\Test\Foo@bar');
    }
}
```

![screen shot 2015-06-11 at 9 09 19 pm](https://cloud.githubusercontent.com/assets/172966/8107739/2c5ad998-107e-11e5-9c2c-eb393f0e7bb2.png)

crynobone replied:
And this works too.

```php
$router->group(['namespace' => 'App\Test'], function ($router) {
    $router->get('foo', 'Foo@bar');
});
```

rigor789 replied:
Here's a gist of the issue: https://gist.github.com/rigor789/aa0bd84c4028c629a99d

Note: It's a fresh laravel install.

crynobone replied:
I assume the `routes.php` is `app/Http/routes.php` it `RouteServiceProvider.php` still contain

```php
    public function map(Router $router)
    {
        $router->group(['namespace' => $this->namespace], function ($router) {
            require app_path('Http/routes.php');
        });
    }
```

Everything under `app/Http/routes.php` would append the root namespace because it is wrap in `$router->group()`.

rigor789 replied:
I am aware the ServiceProvider adds the default namespace, but if I prefix it with a backslash it defaults back to the global namespace as intended. If that works, why can't / why shouldn't ```action()``` work that way as well?

crynobone replied:
> but if I prefix it with a backslash it defaults back to the global namespace as intended. If that works

I would argue that this can be consider as a bug (but that just me). Everything under a namespaced `Route::group` should append the suggested namespace.

And even if you want the framework to support such option, deleting a testcase that test something else instead of creating new testcase to support your suggestion and the same time doesn't break existing code is the way to go.

rigor789 replied:
I personally use named routes for everything, this issue doesn't affect me, I just want to eliminate small "bugs" like these for people who would eventually use them in such scenario.

You are right, I should've created a new test case, but if I did, one of the two would always fail.

crynobone replied:
> I just want to eliminate small "bugs" like these for people who would eventually use them in such scenario.

> but if I did, one of the two would always fail.

Which in the process would break other people code (actual running code, since it been the structure from `v5.0.0` till today).

jack-webster replied:
> if I did, one of the two would always fail.

From my brief testing, I don't think that's true. I think if you create a new test case additionally, and apply your fix, then both test cases pass..

GrahamCampbell replied:
Then actually do that then...

rigor789 replied:
Submitted a new pull request, please suggest a better fix, or give inputs on the proposed one!

rigor789 commented:
@GrahamCampbell I would definitely consider it a bug. If for some reason you have a controller that isn't under the ```App\Http\Conrollers``` namespace, you can define a route for it
```PHP
Route::get('foo', '\App\Foo\Bar@getFoo')
```
If you go to ```app-domain/foo``` the correct ```getFoo``` method will be called.

However if you want to use the ```action()``` url generator, you should be able to do that as well.


For example:
```PHP
action('\\App\Foo\Bar@getFoo')
```

If you use

```PHP
action('App\Foo\Bar@getFoo')
```
it will look in:
```PHP
App\Http\Controller\App\Foo\Bar
```
Which is expected.

On the other hand if you do
```PHP
action('\\App\Foo\Bar@getFoo')
```
it does look for
```PHP
App\Foo\Bar@getFoo
```
but it can't find the action since

```PHP
\App\Foo\Bar@getFoo != App\Foo\Bar@getFoo
```
caused by trimming the slashes.

GrahamCampbell commented:
You have modified a test to change the implementation from correct to incorrect.

jack-webster commented:
@GrahamCampbell The test wasn't sufficient to check this scenario..

rigor789 commented:
Could you please elaboorate how is it incorrect? @GrahamCampbell

RomainLanz commented:
You need to add test and not remove older one!

rigor789 committed:
Updated the test to confirm the url action generator bug.

By default laravel uses the root namespace of "App\Http\Controllers", to have a controller in another namespace we have to use "\My\Other\Namespace\Controller@foo".

The test wasn't adding the extra backslash, thus giving a false positive result.

rigor789 committed:
Proposed fix for #9206

jack-webster commented:
I've looked into this, and I think @rigor789 is right, the original test behaviour doesn't seem to make sense, BUT I can undersand that it's undesirable to remove it and break working code.

I have created my own commits with an additional test, and a modification to UrlGenerator that satisfies both tests.
